### PR TITLE
Update the legalization passes.

### PR DESCRIPTION
**Pull request by s-perron**
Changes the set of optimizations done for legalization.  While doing
this, I added documentation to explain why we want each optimization.

A new option "--legalize-hlsl" is added so the legalization passes can
be easily run from the command line.

Implements issue #1118.

### CODE REVIEW

File: source/opt/optimizer.cpp [hunk 1]
```diff
@@ -66,19 +66,50 @@ Optimizer& Optimizer::RegisterPass(PassToken&& p) {
...
+          // scalar replacement.
+          .RegisterPass(CreateInsertExtractElimPass())
+          // TODO: Add constant propagation here
+          // May need loop unrolling here see
+          // https://github.com/Microsoft/DirectXShaderCompiler/pull/930
+          .RegisterPass(CreateCFGCleanupPass())
+          // Get rid of unused code that leave traces of the illegal code.
+          .RegisterPass(CreateAggressiveDCEPass())
+          // TODO: Remove this once ADCE can do it.
+          .RegisterPass(CreateDeadVariableEliminationPass());
```
Comment (antiagainst) [inline source/opt/optimizer.cpp]: Can we have the auxiliary `CreateCompactIdsPass()` here? Running this recipe bloats the result-id ranges used.

File: source/opt/optimizer.cpp [hunk 2]
```diff
@@ -66,19 +66,50 @@ Optimizer& Optimizer::RegisterPass(PassToken&& p) {
   return *this;
 }
 
+// The legalization passes take a spir-v shader generated by an HLSL front-end
+// and turn it into a valid vulkan spir-v shader.  There are two ways in which
+// the code will be invalid at the start:
+//
+// 1) There will of opaque objects, like images, which will be passed around
```
Comment (antiagainst) [inline source/opt/optimizer.cpp]: s/of/be/

File: source/opt/optimizer.cpp (hunk 1 above)
Comment (dneto0) [reply source/opt/optimizer.cpp]: But ID compaction is not strictly a legalization.  If you think it's beneficial for other reasons then you should apply it separately. ??

Comment (antiagainst) [reply source/opt/optimizer.cpp]: That's true. I can register it separately then.

File: tools/opt/opt.cpp [hunk 3]
```diff
@@ -421,6 +434,8 @@ OptStatus ParseFlags(int argc, const char** argv, Optimizer* optimizer,
         optimizer->RegisterPerformancePasses();
       } else if (0 == strcmp(cur_arg, "-Os")) {
         optimizer->RegisterSizePasses();
+      } else if (0 == strcmp(cur_arg, "-legalize-hlsl")) {
```
Comment (antiagainst) [inline tools/opt/opt.cpp]: Should use double dash here.

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as 7505d24225

**Pull request closed without merging**